File: some-python/circular-buffer.py
class CircularBuffer:
	def __init__(self, capacity: int = 8):
		if capacity <= 0 or (capacity & (capacity - 1)) != 0:
			raise ValueError("Capacity must be a power of 2")

		self.capacity = capacity
		self.buffer = [None] * capacity
		self.head = 0
		self.tail = 0
		self.count = 0

	def __str__(self) -> str:
		items = [
			self.buffer[(self.tail +i) & (self.capacity - 1)]
			for i in range(self.count)
		]

		return f"CircularBuffer({items}, count={self.count}/{self.capacity})"

	def __repr__(self) -> str:
		return self.__str__()

	def is_empty(self) -> bool:
		return self.count == 0

	def is_full(self) -> bool:
		return self.count == self.capacity

	def push(self, item) -> bool:
		"""Push an item & return false if full"""
		if self.is_full():
			return False

		self.buffer[self.head] = item
		self.head = (self.head + 1) & (self.capacity - 1)
		self.count += 1

		return True
	
	def pop(self):
		if self.is_empty():
			return False

		item = self.buffer[self.tail]
		self.buffer[self.tail] = None
		self.tail = (self.tail + 1) & (self.capacity - 1)
		self.count -= 1
		return item

	


cb = CircularBuffer();
cb.push(10)
cb.push(12)
cb.push(22)
cb.pop()
cb.push(24)
print(cb)

File: some-rust/circular-buffer.rs
#[derive(Debug)]

pub struct CircularBuffer<T, const N:usize> {
	buffer: [Option<T>; N],
	head: usize,
	tail: usize,
	count: usize,
}

impl<T, const N: usize> CircularBuffer<T, N> {
	pub fn new() -> Self {
		assert!(N > 0 && (N & (N - 1) == 0), "Capacity N must be a power of 2");

		Self {
			buffer: std::array::from_fn(|_| None),
			head: 0,
			tail: 0,
			count: 0,
		}
	}

	pub fn is_full(&self) -> bool {
		self.count == N
	}

	pub fn is_empty(&self) -> bool {
		self.count == 0
	}

	pub fn push(&mut self, item: T) -> Result<(), T> {
		if self.is_full() {
			return Err(item);
		}

		self.buffer[self.head] = Some(item);
		self.head = (self.head + 1) & (N - 1);

		self.count += 1;
		Ok(())
	}

	pub fn pop(&mut self) -> Option<T> {
		if self.is_empty() {
			return None;
		}

		let item = self.buffer[self.tail].take();
		self.tail = (self.tail + 1) & (N - 1);
		self.count -= 1;

		item
	}
}

fn main() {
	let mut cb = CircularBuffer::<i32, 8>::new();

	cb.push(12).unwrap();
	cb.push(24).unwrap();
	cb.push(32).unwrap();

	while let Some(val) = cb.pop() {
		println!("current: {:?}", cb);
		println!("Popped: {}", val);
	}
}

